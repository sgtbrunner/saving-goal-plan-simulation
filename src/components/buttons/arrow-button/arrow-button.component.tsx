import { ArrowButtonProps } from './arrow-button.types';
import Arrow from '../../../assets/icons/arrow.svg';
import { Button } from './arrow-button.styles';

const ArrowButton = ({
  direction = 'right',
  disabled = false,
}: ArrowButtonProps): JSX.Element => (
  <Button direction={direction} disabled={disabled} type="button">
    <img src={Arrow} alt={`arrow-${direction}`} />
  </Button>
);

export default ArrowButton;
