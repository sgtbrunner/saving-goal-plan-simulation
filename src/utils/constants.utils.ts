// OPERATIONAL CONSTANTS
export const FIRST_STEP = 1;
export const LOCALE = 'en';
export const INITIAL_GOAL_AMOUNT = 25000;

// GENERAL CONSTANTS
export const ALT = 'alt';
export const BANNER = 'banner';
export const BUTTON = 'button';
export const ID = 'id';
export const IMG = 'img';
export const ORIGIN = 'Origin';
export const LEFT = 'left';
export const RIGHT = 'right';
export const TEST = 'test';
export const WHITE = 'white';
