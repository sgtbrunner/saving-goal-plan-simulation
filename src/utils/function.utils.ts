export const clearInvalidCharacters = (value: string): string =>
  value.replace(/\D|^0+/g, '');

export const convertInputToNumber = (value: string): number => {
  if (value.includes('.')) {
    const [decimalPart, fractionalPart] = value.split('.');
    return Number(
      [
        clearInvalidCharacters(decimalPart),
        '.',
        clearInvalidCharacters(fractionalPart),
      ].join('')
    );
  }

  return Number(value.replace(/\D|^0+/g, ''));
};
