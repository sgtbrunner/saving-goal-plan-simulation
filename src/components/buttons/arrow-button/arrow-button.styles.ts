import styled, { css, FlattenSimpleInterpolation } from 'styled-components';

import { ArrowButtonProps } from './arrow-button.types';

const getArrowDirection = ({
  direction,
}: ArrowButtonProps): FlattenSimpleInterpolation => css`
  transform: rotate(${direction === 'right' ? '0deg' : '180deg'});
`;

export const Button = styled.button`
  padding: 20px;
  border: none;
  background-color: transparent;
  color: var(--blue-gray-300-color);
  ${getArrowDirection}
`;
