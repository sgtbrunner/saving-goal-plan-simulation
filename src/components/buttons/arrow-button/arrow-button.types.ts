import { LEFT, RIGHT } from '../../../utils/constants.utils';

export type ArrowButtonProps = {
  id: string;
  direction?: typeof LEFT | typeof RIGHT;
  disabled?: boolean;
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
};
