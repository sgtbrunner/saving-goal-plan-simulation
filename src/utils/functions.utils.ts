import { FIRST_STEP, LOCALE } from './constants.utils';
import { MonthPickerInputState } from './types.utils';

export const getDate = (customDate?: string | number): Date =>
  customDate ? new Date(customDate) : new Date();

export const monthFormatter = new Intl.DateTimeFormat(LOCALE, {
  month: 'long',
});

export const yearFormatter = new Intl.DateTimeFormat(LOCALE, {
  year: 'numeric',
});

export const getMonthPickerState = (
  step = FIRST_STEP,
  modifier = 0
): MonthPickerInputState => {
  const today = getDate();
  const targetMonthISO = today.setMonth(today.getMonth() + step + modifier, 1);

  return {
    step: step + modifier,
    year: yearFormatter.format(targetMonthISO),
    month: monthFormatter.format(targetMonthISO),
  };
};

export const getSignificantAndFractionalDigits = (
  value: string
): Array<string> => value.split('.');
