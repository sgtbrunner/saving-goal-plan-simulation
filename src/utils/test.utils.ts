import { INITIAL_GOAL_AMOUNT } from './constants.utils';
import { getDate, monthFormatter, yearFormatter } from './functions.utils';
import { SavingGoalState } from './types.utils';

// TAGS
export const ALT = 'alt';
export const BANNER = 'banner';
export const BUTTON = 'button';
export const IMG = 'img';
export const TEST = 'test';

// TEST IDs
export const CARD = 'card';
export const MAIN_COMPONENT = 'main-component';
export const MAIN_TITLE = 'main-title';
export const MONTHLY_AMOUNT = 'monthly-amount';
export const SUMMARY_TEXT = 'summary-text';

// FUNCTIONS
export const getDefaultSavingGoal = (): SavingGoalState => {
  const today = getDate();
  const nextMonth = today.setMonth(today.getMonth() + 1, 1);
  return {
    goalAmount: INITIAL_GOAL_AMOUNT,
    reachDateYear: yearFormatter.format(nextMonth),
    reachDateMonth: monthFormatter.format(nextMonth),
    monthlyAmount: INITIAL_GOAL_AMOUNT,
    monthlyDeposits: 1,
  };
};
