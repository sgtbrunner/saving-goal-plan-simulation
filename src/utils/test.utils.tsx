import React from 'react';
import { render, RenderResult } from '@testing-library/react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';

import rootReducer from '../redux/root-reducer';
import { INITIAL_GOAL_AMOUNT } from './constants.utils';
import { getDate, monthFormatter, yearFormatter } from './functions.utils';
import { SavingGoalState } from './types.utils';

// TAG CONSTANTS
export const ALT = 'alt';
export const BANNER = 'banner';
export const BUTTON = 'button';
export const IMG = 'img';
export const TEST = 'test';
export const AMOUNT = 'amount';

// TEST ID CONSTANTS
export const CARD = 'card';
export const MAIN_COMPONENT = 'main-component';
export const MAIN_TITLE = 'main-title';
export const MONTHLY_AMOUNT = 'monthly-amount';
export const SUMMARY_TEXT = 'summary-text';
export const CURRENCY_INPUT = 'currency-input';

// FUNCTIONS
export const getDefaultSavingGoal = (): SavingGoalState => {
  const today = getDate();
  const nextMonth = today.setMonth(today.getMonth() + 1, 1);
  return {
    goalAmount: INITIAL_GOAL_AMOUNT,
    reachDateYear: yearFormatter.format(nextMonth),
    reachDateMonth: monthFormatter.format(nextMonth),
    monthlyAmount: INITIAL_GOAL_AMOUNT,
    monthlyDeposits: 1,
  };
};

// RENDERER
type WrapperPropsType = {
  children: React.ReactNode | React.ReactNode[];
};

export const renderWithRedux = (
  component: React.ReactElement,
  store = createStore(rootReducer)
): RenderResult => {
  const ComponentWrapper = (props: WrapperPropsType) => (
    <Provider store={store}>{props.children}</Provider>
  );

  return render(component, {
    wrapper: ComponentWrapper as React.ComponentType,
  });
};
