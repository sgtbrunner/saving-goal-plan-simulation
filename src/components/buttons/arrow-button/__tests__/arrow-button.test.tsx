import { screen, render, RenderResult, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { ID, LEFT } from '../../../../utils/constants.utils';
import { ALT, BUTTON, IMG, TEST } from '../../../../utils/test.utils';
import { ArrowButtonProps } from '../arrow-button.types';
import ArrowButton from '../arrow-button.component';

const defaultProps: ArrowButtonProps = {
  id: TEST,
  onClick: jest.fn(),
};

const mountComponent = (newProps?: ArrowButtonProps): RenderResult =>
  render(<ArrowButton {...defaultProps} {...newProps} />);

describe('ArrowButton component', () => {
  it('should match snapshot', () => {
    const { container } = mountComponent();

    expect(container).toMatchSnapshot();
  });

  it(`should render a right oriented arrow button if the default props only are provided`, () => {
    mountComponent();

    const arrowButton = screen.getByRole(BUTTON);
    expect(arrowButton).toBeInTheDocument();
    expect(arrowButton).toHaveProperty(ID, defaultProps.id);

    const buttonImage = within(arrowButton).getByRole(IMG);
    expect(buttonImage).toBeInTheDocument();
    expect(buttonImage).toHaveProperty(ALT, 'arrow-right');
  });

  it('should render a left oriented arrow button if the proper props are provided', () => {
    const newProps: ArrowButtonProps = {
      id: 'new-test',
      direction: LEFT,
    };
    mountComponent(newProps);

    const arrowButton = screen.getByRole(BUTTON);
    expect(arrowButton).toBeInTheDocument();
    expect(arrowButton).toHaveProperty(ID, newProps.id);

    const buttonImage = within(arrowButton).getByRole(IMG);
    expect(buttonImage).toBeInTheDocument();
    expect(buttonImage).toHaveProperty(ALT, 'arrow-left');
  });

  it('should call onClick function if button IS NOT disabled', () => {
    mountComponent();

    const arrowButton = screen.getByRole(BUTTON);
    userEvent.click(arrowButton);
    expect(defaultProps.onClick).toHaveBeenCalledTimes(1);
  });

  it('should NOT call onClick function if button IS disabled', () => {
    const newProps: ArrowButtonProps = {
      id: defaultProps.id,
      disabled: true,
    };
    mountComponent(newProps);

    const arrowButton = screen.getByRole(BUTTON);
    userEvent.click(arrowButton);
    expect(defaultProps.onClick).not.toHaveBeenCalled();
  });
});
