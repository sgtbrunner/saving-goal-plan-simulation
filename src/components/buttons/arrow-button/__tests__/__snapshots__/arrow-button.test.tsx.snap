// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArrowButton component should match snapshot 1`] = `
<div>
  <button
    class="sc-bdnxRM leqYLS"
    id="test"
    type="button"
  >
    <img
      alt="arrow-right"
      src="arrow-right.svg"
    />
  </button>
</div>
`;
